tools/foxglove/README.md

```diff
@@ -0,0 +1,45 @@
+# Foxglove Converter
+
+[Foxglove Studio](https://foxglove.dev/download) is a tool to visualize robotics data. We've written a script to convert openpilot log data into mcap files for use inside foxglove studio.
+
+## Installation
+
+Ensure you've [set up the openpilot environment](../README.md), once the environment is setup the converter is ready to run.
+
+## Usage
+
+```
+$ ./cereal_to_mcap.py -h
+usage: cereal_to_mcap.py [-h] [--demo] [--qlog] [--ci] [--fcam] [--dcam] [--ecam] [route_or_segment_name] [segment_count]
+
+A helper to convert openpilot routes to mcap files for foxglove studio
+
+positional arguments:
+  route_or_segment_name
+                        The route or segment name to plot (cabana share URL accepted) (default: None)
+  segment_count         The number of segments to plot (default: None)
+
+options:
+  -h, --help            show this help message and exit
+  --demo                Use the demo route instead of providing one (default: False)
+  --qlog                Use qlogs (default: False)
+  --ci                  Download data from openpilot CI bucket (default: False)
+  --fcam                Include fcamera data (default: False)
+  --dcam                Include dcamera data (default: False)
+  --ecam                Include ecamera data (default: False)
+
+```
+
+Examples using route name:
+
+`./cereal_to_mcap.py "a2a0ccea32023010|2023-07-27--13-01-19"`
+
+Examples using segment name:
+
+`./cereal_to_mcap.py "a2a0ccea32023010|2023-07-27--13-01-19--1"`
+
+## Demo
+
+For a quick demo, go through the installation step and run this command:
+
+`./cereal_to_mcap.py --demo --qlog`
```

Comment:
this didn't work on my machine
```
(openpilot-py3.11) bash-3.2$ ./cereal_to_mcap.py --demo --qlog
warning, foxglove does not support lists in lists, skipping field: liveEventDEPRECATED
warning, foxglove does not support lists in lists, skipping field: ethernetDataDEPRECATED
warning, foxglove does not support lists in lists, skipping field: cellInfoDEPRECATED
warning, foxglove does not support lists in lists, skipping field: wifiScanDEPRECATED
warning, foxglove does not support lists in lists, skipping field: trafficEventsDEPRECATED
warning, foxglove does not support lists in lists, skipping field: orbObservationDEPRECATED
warning, foxglove does not support lists in lists, skipping field: x
warning, foxglove does not support lists in lists, skipping field: points
Loading Segment(s)


warning, foxglove does not support lists in lists, skipping field: liveEventDEPRECATED
warning, foxglove does not support lists in lists, skipping field: ethernetDataDEPRECATED
warning, foxglove does not support lists in lists, skipping field: cellInfoDEPRECATED
warning, foxglove does not support lists in lists, skipping field: wifiScanDEPRECATED
warning, foxglove does not support lists in lists, skipping field: trafficEventsDEPRECATED
warning, foxglove does not support lists in lists, skipping field: orbObservationDEPRECATED
warning, foxglove does not support lists in lists, skipping field: x
warning, foxglove does not support lists in lists, skipping field: liveEventDEPRECATED
warning, foxglove does not support lists in lists, skipping field: ethernetDataDEPRECATED
warning, foxglove does not support lists in lists, skipping field: cellInfoDEPRECATED
warning, foxglove does not support lists in lists, skipping field: wifiScanDEPRECATED
warning, foxglove does not support lists in lists, skipping field: trafficEventsDEPRECATED
warning, foxglove does not support lists in lists, skipping field: orbObservationDEPRECATED
warning, foxglove does not support lists in lists, skipping field: x
warning, foxglove does not support lists in lists, skipping field: points
warning, foxglove does not support lists in lists, skipping field: liveEventDEPRECATED
warning, foxglove does not support lists in lists, skipping field: ethernetDataDEPRECATED
warning, foxglove does not support lists in lists, skipping field: cellInfoDEPRECATED
warning, foxglove does not support lists in lists, skipping field: wifiScanDEPRECATED
warning, foxglove does not support lists in lists, skipping field: trafficEventsDEPRECATED
warning, foxglove does not support lists in lists, skipping field: orbObservationDEPRECATED
warning, foxglove does not support lists in lists, skipping field: points
warning, foxglove does not support lists in lists, skipping field: liveEventDEPRECATED
warning, foxglove does not support lists in lists, skipping field: x
warning, foxglove does not support lists in lists, skipping field: ethernetDataDEPRECATED
warning, foxglove does not support lists in lists, skipping field: cellInfoDEPRECATED
warning, foxglove does not support lists in lists, skipping field: wifiScanDEPRECATED
warning, foxglove does not support lists in lists, skipping field: liveEventDEPRECATED
warning, foxglove does not support lists in lists, skipping field: ethernetDataDEPRECATED
warning, foxglove does not support lists in lists, skipping field: cellInfoDEPRECATED
warning, foxglove does not support lists in lists, skipping field: wifiScanDEPRECATED
warning, foxglove does not support lists in lists, skipping field: trafficEventsDEPRECATED
warning, foxglove does not support lists in lists, skipping field: orbObservationDEPRECATED
warning, foxglove does not support lists in lists, skipping field: x
warning, foxglove does not support lists in lists, skipping field: liveEventDEPRECATED
warning, foxglove does not support lists in lists, skipping field: ethernetDataDEPRECATED
warning, foxglove does not support lists in lists, skipping field: cellInfoDEPRECATED
warning, foxglove does not support lists in lists, skipping field: wifiScanDEPRECATED
warning, foxglove does not support lists in lists, skipping field: trafficEventsDEPRECATED
warning, foxglove does not support lists in lists, skipping field: orbObservationDEPRECATED
warning, foxglove does not support lists in lists, skipping field: trafficEventsDEPRECATED
warning, foxglove does not support lists in lists, skipping field: orbObservationDEPRECATED
warning, foxglove does not support lists in lists, skipping field: x
warning, foxglove does not support lists in lists, skipping field: points
warning, foxglove does not support lists in lists, skipping field: liveEventDEPRECATED
warning, foxglove does not support lists in lists, skipping field: ethernetDataDEPRECATED
warning, foxglove does not support lists in lists, skipping field: cellInfoDEPRECATED
warning, foxglove does not support lists in lists, skipping field: wifiScanDEPRECATED
warning, foxglove does not support lists in lists, skipping field: x
warning, foxglove does not support lists in lists, skipping field: trafficEventsDEPRECATED
warning, foxglove does not support lists in lists, skipping field: orbObservationDEPRECATED
warning, foxglove does not support lists in lists, skipping field: x
warning, foxglove does not support lists in lists, skipping field: points
warning, foxglove does not support lists in lists, skipping field: points
warning, foxglove does not support lists in lists, skipping field: points
warning, foxglove does not support lists in lists, skipping field: liveEventDEPRECATED
warning, foxglove does not support lists in lists, skipping field: liveEventDEPRECATED
warning, foxglove does not support lists in lists, skipping field: ethernetDataDEPRECATED
warning, foxglove does not support lists in lists, skipping field: cellInfoDEPRECATED
warning, foxglove does not support lists in lists, skipping field: wifiScanDEPRECATED
warning, foxglove does not support lists in lists, skipping field: liveEventDEPRECATED
warning, foxglove does not support lists in lists, skipping field: ethernetDataDEPRECATED
warning, foxglove does not support lists in lists, skipping field: cellInfoDEPRECATED
warning, foxglove does not support lists in lists, skipping field: wifiScanDEPRECATED
warning, foxglove does not support lists in lists, skipping field: points
warning, foxglove does not support lists in lists, skipping field: liveEventDEPRECATED
warning, foxglove does not support lists in lists, skipping field: ethernetDataDEPRECATED
warning, foxglove does not support lists in lists, skipping field: cellInfoDEPRECATED
warning, foxglove does not support lists in lists, skipping field: wifiScanDEPRECATED
warning, foxglove does not support lists in lists, skipping field: trafficEventsDEPRECATED
warning, foxglove does not support lists in lists, skipping field: orbObservationDEPRECATED
warning, foxglove does not support lists in lists, skipping field: x
warning, foxglove does not support lists in lists, skipping field: liveEventDEPRECATED
warning, foxglove does not support lists in lists, skipping field: points
warning, foxglove does not support lists in lists, skipping field: ethernetDataDEPRECATED
warning, foxglove does not support lists in lists, skipping field: cellInfoDEPRECATED
warning, foxglove does not support lists in lists, skipping field: wifiScanDEPRECATED
warning, foxglove does not support lists in lists, skipping field: trafficEventsDEPRECATED
warning, foxglove does not support lists in lists, skipping field: orbObservationDEPRECATED
warning, foxglove does not support lists in lists, skipping field: x
warning, foxglove does not support lists in lists, skipping field: ethernetDataDEPRECATED
warning, foxglove does not support lists in lists, skipping field: cellInfoDEPRECATED
warning, foxglove does not support lists in lists, skipping field: wifiScanDEPRECATED
warning, foxglove does not support lists in lists, skipping field: trafficEventsDEPRECATED
warning, foxglove does not support lists in lists, skipping field: orbObservationDEPRECATED
warning, foxglove does not support lists in lists, skipping field: x
warning, foxglove does not support lists in lists, skipping field: trafficEventsDEPRECATED
warning, foxglove does not support lists in lists, skipping field: orbObservationDEPRECATED
warning, foxglove does not support lists in lists, skipping field: trafficEventsDEPRECATED
warning, foxglove does not support lists in lists, skipping field: orbObservationDEPRECATED
warning, foxglove does not support lists in lists, skipping field: points
warning, foxglove does not support lists in lists, skipping field: liveEventDEPRECATED
warning, foxglove does not support lists in lists, skipping field: x
warning, foxglove does not support lists in lists, skipping field: ethernetDataDEPRECATED
warning, foxglove does not support lists in lists, skipping field: cellInfoDEPRECATED
warning, foxglove does not support lists in lists, skipping field: x
warning, foxglove does not support lists in lists, skipping field: wifiScanDEPRECATED
warning, foxglove does not support lists in lists, skipping field: points
warning, foxglove does not support lists in lists, skipping field: points
warning, foxglove does not support lists in lists, skipping field: trafficEventsDEPRECATED
warning, foxglove does not support lists in lists, skipping field: orbObservationDEPRECATED
warning, foxglove does not support lists in lists, skipping field: x
warning, foxglove does not support lists in lists, skipping field: points
warning, foxglove does not support lists in lists, skipping field: liveEventDEPRECATED
warning, foxglove does not support lists in lists, skipping field: liveEventDEPRECATED
warning, foxglove does not support lists in lists, skipping field: ethernetDataDEPRECATED
warning, foxglove does not support lists in lists, skipping field: cellInfoDEPRECATED
warning, foxglove does not support lists in lists, skipping field: wifiScanDEPRECATED
warning, foxglove does not support lists in lists, skipping field: trafficEventsDEPRECATED
warning, foxglove does not support lists in lists, skipping field: orbObservationDEPRECATED
warning, foxglove does not support lists in lists, skipping field: x
warning, foxglove does not support lists in lists, skipping field: points
warning, foxglove does not support lists in lists, skipping field: points
warning, foxglove does not support lists in lists, skipping field: liveEventDEPRECATED
warning, foxglove does not support lists in lists, skipping field: ethernetDataDEPRECATED
warning, foxglove does not support lists in lists, skipping field: cellInfoDEPRECATED
warning, foxglove does not support lists in lists, skipping field: wifiScanDEPRECATED
warning, foxglove does not support lists in lists, skipping field: liveEventDEPRECATED
warning, foxglove does not support lists in lists, skipping field: ethernetDataDEPRECATED
warning, foxglove does not support lists in lists, skipping field: cellInfoDEPRECATED
warning, foxglove does not support lists in lists, skipping field: wifiScanDEPRECATED
warning, foxglove does not support lists in lists, skipping field: trafficEventsDEPRECATED
warning, foxglove does not support lists in lists, skipping field: orbObservationDEPRECATED
warning, foxglove does not support lists in lists, skipping field: x
warning, foxglove does not support lists in lists, skipping field: liveEventDEPRECATED
warning, foxglove does not support lists in lists, skipping field: points
warning, foxglove does not support lists in lists, skipping field: ethernetDataDEPRECATED
warning, foxglove does not support lists in lists, skipping field: cellInfoDEPRECATED
warning, foxglove does not support lists in lists, skipping field: wifiScanDEPRECATED
warning, foxglove does not support lists in lists, skipping field: trafficEventsDEPRECATED
warning, foxglove does not support lists in lists, skipping field: orbObservationDEPRECATED
warning, foxglove does not support lists in lists, skipping field: x
warning, foxglove does not support lists in lists, skipping field: points
warning, foxglove does not support lists in lists, skipping field: liveEventDEPRECATED
warning, foxglove does not support lists in lists, skipping field: ethernetDataDEPRECATED
warning, foxglove does not support lists in lists, skipping field: ethernetDataDEPRECATED
warning, foxglove does not support lists in lists, skipping field: cellInfoDEPRECATED
warning, foxglove does not support lists in lists, skipping field: wifiScanDEPRECATED
warning, foxglove does not support lists in lists, skipping field: cellInfoDEPRECATED
warning, foxglove does not support lists in lists, skipping field: wifiScanDEPRECATED
warning, foxglove does not support lists in lists, skipping field: trafficEventsDEPRECATED
warning, foxglove does not support lists in lists, skipping field: orbObservationDEPRECATED
warning, foxglove does not support lists in lists, skipping field: x
warning, foxglove does not support lists in lists, skipping field: trafficEventsDEPRECATED
warning, foxglove does not support lists in lists, skipping field: orbObservationDEPRECATED
warning, foxglove does not support lists in lists, skipping field: liveEventDEPRECATED
warning, foxglove does not support lists in lists, skipping field: ethernetDataDEPRECATED
warning, foxglove does not support lists in lists, skipping field: cellInfoDEPRECATED
warning, foxglove does not support lists in lists, skipping field: wifiScanDEPRECATED
warning, foxglove does not support lists in lists, skipping field: x
warning, foxglove does not support lists in lists, skipping field: trafficEventsDEPRECATED
warning, foxglove does not support lists in lists, skipping field: orbObservationDEPRECATED
warning, foxglove does not support lists in lists, skipping field: x
warning, foxglove does not support lists in lists, skipping field: points
warning, foxglove does not support lists in lists, skipping field: liveEventDEPRECATED
warning, foxglove does not support lists in lists, skipping field: ethernetDataDEPRECATED
warning, foxglove does not support lists in lists, skipping field: trafficEventsDEPRECATED
warning, foxglove does not support lists in lists, skipping field: orbObservationDEPRECATED
warning, foxglove does not support lists in lists, skipping field: liveEventDEPRECATED
warning, foxglove does not support lists in lists, skipping field: ethernetDataDEPRECATED
warning, foxglove does not support lists in lists, skipping field: cellInfoDEPRECATED
warning, foxglove does not support lists in lists, skipping field: wifiScanDEPRECATED
warning, foxglove does not support lists in lists, skipping field: cellInfoDEPRECATED
warning, foxglove does not support lists in lists, skipping field: wifiScanDEPRECATED
warning, foxglove does not support lists in lists, skipping field: trafficEventsDEPRECATED
warning, foxglove does not support lists in lists, skipping field: orbObservationDEPRECATED
warning, foxglove does not support lists in lists, skipping field: x
warning, foxglove does not support lists in lists, skipping field: x
warning, foxglove does not support lists in lists, skipping field: points
warning, foxglove does not support lists in lists, skipping field: trafficEventsDEPRECATED
warning, foxglove does not support lists in lists, skipping field: orbObservationDEPRECATED
warning, foxglove does not support lists in lists, skipping field: liveEventDEPRECATED
warning, foxglove does not support lists in lists, skipping field: points
warning, foxglove does not support lists in lists, skipping field: x
warning, foxglove does not support lists in lists, skipping field: ethernetDataDEPRECATED
warning, foxglove does not support lists in lists, skipping field: cellInfoDEPRECATED
warning, foxglove does not support lists in lists, skipping field: wifiScanDEPRECATED
warning, foxglove does not support lists in lists, skipping field: trafficEventsDEPRECATED
warning, foxglove does not support lists in lists, skipping field: orbObservationDEPRECATED
warning, foxglove does not support lists in lists, skipping field: x
warning, foxglove does not support lists in lists, skipping field: points
warning, foxglove does not support lists in lists, skipping field: liveEventDEPRECATED
warning, foxglove does not support lists in lists, skipping field: ethernetDataDEPRECATED
warning, foxglove does not support lists in lists, skipping field: cellInfoDEPRECATED
warning, foxglove does not support lists in lists, skipping field: wifiScanDEPRECATED
warning, foxglove does not support lists in lists, skipping field: trafficEventsDEPRECATED
warning, foxglove does not support lists in lists, skipping field: orbObservationDEPRECATED
warning, foxglove does not support lists in lists, skipping field: x
warning, foxglove does not support lists in lists, skipping field: points
warning, foxglove does not support lists in lists, skipping field: points
warning, foxglove does not support lists in lists, skipping field: points
warning, foxglove does not support lists in lists, skipping field: points



Segment(s) Loaded
Starting Conversion
Percent converted: 0.07%Traceback (most recent call last):
  File "/Users/adeebshihadeh/openpilot/tools/foxglove/./cereal_to_mcap.py", line 199, in <module>
    download_and_convert_route(route_or_segment_name, args.segment_count, args.qlog, args.ci, args.fcam, args.dcam, args.ecam)
  File "/Users/adeebshihadeh/openpilot/tools/foxglove/./cereal_to_mcap.py", line 94, in download_and_convert_route
    convert_log(f"{str(route_or_segment_name.canonical_name).split('|')[1]}.mcap", tmp.name, cams)
  File "/Users/adeebshihadeh/openpilot/tools/foxglove/./cereal_to_mcap.py", line 172, in convert_log
    message(writer, typeToChannel[str(event.which)], e, offset, e)
  File "/Users/adeebshihadeh/openpilot/openpilot/tools/foxglove/utils.py", line 26, in message
    writer.add_message(
  File "/Users/adeebshihadeh/Library/Caches/pypoetry/virtualenvs/openpilot-5VL5NVLP-py3.11/lib/python3.11/site-packages/mcap/writer.py", line 196, in add_message
    self.__chunk_builder.add_message(message)
  File "/Users/adeebshihadeh/Library/Caches/pypoetry/virtualenvs/openpilot-5VL5NVLP-py3.11/lib/python3.11/site-packages/mcap/_chunk_builder.py", line 45, in add_message
    message.write(self.record_writer)
  File "/Users/adeebshihadeh/Library/Caches/pypoetry/virtualenvs/openpilot-5VL5NVLP-py3.11/lib/python3.11/site-packages/mcap/records.py", line 312, in write
    stream.write8(self.log_time)
  File "/Users/adeebshihadeh/Library/Caches/pypoetry/virtualenvs/openpilot-5VL5NVLP-py3.11/lib/python3.11/site-packages/mcap/data_stream.py", line 104, in write8
    self.write(struct.pack("<Q", value))
               ^^^^^^^^^^^^^^^^^^^^^^^^
struct.error: argument out of range
(openpilot-py3.11) bash-3.2$ 
```

Reply:
i just fixed the demo route. had to do with me calculating the timestamp using the difference between wallTimeMonos and the logMonoTime. I'm just ignoring the offset now if wallTimeMonos is 0, but that does result in the demo route having incorrect timestamps in foxglove.

What I don't understand is why the warning logs repeated like that. The part of code that prints those warnings out only runs once at the beginning of the script. Once it starts loading the segments it shouldn't be possible to see those warnings printed out anymore. It's almost like you had multiple instances of the script running